### PR TITLE
Add AVMaxFileSize config option

### PR DESCRIPTION
**Pull request by VicDeo**
### Description
- Adds File size limit
- Fixes https://github.com/owncloud/files_antivirus/issues/124 and https://github.com/owncloud/files_antivirus/issues/125

### Changes in Docs:
__Stream Length__ (bytes)  `26214400` (25Mb) by default.  __Should be <=__  than  `StreamMaxLength` ClamAV value
 __File size limit, -1 means no limit__ (bytes)  `-1` by default - prevent background scanner from checking large files (Has no effect while scanning newly uploaded files)

### To test: 
1. set StreamMaxLength in `/etc/clamd.conf` to e.g. 10M 
2. Configure the app in daemon or daemon(socket) mode
3. cat several 100Mb files with EICAR signature
4. Try upload or background scan

### CODE REVIEW

**Comment by VicDeo** [comment]
@PVince81 please have a look

File: lib/backgroundscanner.php [hunk 1]
```diff
@@ -133,6 +93,68 @@ public function run(){
...
+			->leftJoin('fc', 'files_antivirus', 'fa', $qb->expr()->eq('fa.fileid', 'fc.fileid'))
+			->innerJoin(
+				'fc',
+				'storages',
+				'ss',
+				$qb->expr()->andX(
+					$qb->expr()->eq('fc.storage', 'ss.numeric_id'),
+					$qb->expr()->orX(
+						$qb->expr()->like('ss.id', $qb->expr()->literal('local::%')),
+						$qb->expr()->like('ss.id', $qb->expr()->literal('home::%'))
```
Comment (PVince81) [inline lib/backgroundscanner.php]: unrelated, but reminded me of a potential issue, raised as https://github.com/owncloud/files_antivirus/issues/137

**Comment by PVince81** [comment]
Code looks good.

@owncloud/qa can you help testing ?

@VicDeo Any chance to add unit tests of some sorts ? That quite some complex logic there.
Even a test that only covers the stream stuff using a dummy stream would be nice.

For the docs change, please raise a ticket in the documentation repo.

**Comment by PVince81** [comment]
@VicDeo question: what if the virus signature overlaps with the retry-boundary, can it still be detected properly ?

**Comment by PVince81** [comment]
maybe the unit test could have its own dummy signature and have the mock stream receiver detect it (which simulates whatever clamd does)

**Comment by VicDeo** [comment]
@PVince81 
 > what if the virus signature overlaps with the retry-boundary, can it still be detected properly ?

E.g.  When Stream Length contains 3  chunks. And file has 7.
The scanning will be done as
1st scan request: chunks 1-2-3 
2nd scan request: chunks  3-4-5
3rd scan request: chunks 5-6-7

if a signature is spread across chunks 2-3-4 - the answer is `No`
In general ClamAV doesn't support large files at all. So it is still an improvement :)

Hm. I've just came upon an idea to hide chunk size from UI and make it completely internal.

**Comment by VicDeo** [comment]
@PVince81 I added a dumb avir server but it fails to scan more than one file unfortunately

**Comment by sidey79** [comment]
With Last PR, Chunk Size can be removed from the docs i think.

**Comment by VicDeo** [comment]
@sidey79 thanks, updated

**Comment by ManfredSteiner** [comment]
@VicDeo I test branch fix-125 (b3480aef4067f09756428f09b94a8698b135ecd8).

The *byteCount* should be cleared to zero, if the stream is closed ([scanner.php#L143](https://github.com/owncloud/files_antivirus/blob/b3480aef4067f09756428f09b94a8698b135ecd8/lib/scanner.php#L143)). Test with EICAR Virus Signature file results in *Unsupported Media Type*. File not uploaded.

**Comment by ms270169** [comment]
(Conversation before done in #125)
@VicDeo I have tested branch fix-125 (b3480ae).

The byteCount should be cleared to zero, if the stream is closed ([scanner.php#L143](https://github.com/owncloud/files_antivirus/blob/b3480aef4067f09756428f09b94a8698b135ecd8/lib/scanner.php#L143)). Test with EICAR Virus Signature file results in Unsupported Media Type. File not uploaded.

**Comment by VicDeo** [comment]
@ms270169  byteCount is reset to 0 in  `Scanner::initScanner` that is called just one line below

[scanner.php#L144](https://github.com/owncloud/files_antivirus/blob/b3480aef4067f09756428f09b94a8698b135ecd8/lib/scanner.php#L144)

**Comment by ms270169** [comment]
@VicDeo, you are right (I was wrong), byteCount ist cleared to zero in *initScanner()*. Virus scanning works now as desired.

I tested it with the following configuration:
* testfile: 30000 bytes length, 68 byte EICAR virus signature on position 16360-17027
* config-value *Chunk Size*: 8192 (config not used, because I add the file with plus button)
* used chunk-size: 8192
* config-value *Stream Length*: 20000
* config-value *File size limit*: -1

Virus signature overlaps second and third chunk. Virus is only detected if second chunk is repeated after reinitialisation of stream (if block overlapping is implemented).

**Result:**
Virus detection successful, File not accepted by the cloud-system.

The user gets the info *"Unsupported media Type"*, which is a confusing message. If the other bug, leading to this wrong message could not be fixed (see https://github.com/owncloud/files_antivirus/issues/125#issuecomment-259970947), I would recommend to add a log entry in case of virus detection (for example before [avirwrapper.php#L99](https://github.com/owncloud/files_antivirus/blob/2c3e6ec4407a5f94d201c4bbbe1f31b7c9b92c8a/lib/avirwrapper.php#L99)).

**Comment by VicDeo** [comment]
@ms270169
InvalidContentExeption is mapped tonto  HTTP status 415.  "Unsupported media Type" is a text representation for HTTP status 415.
There is no special HTTP status "User just uploaded a virus" unfortunately.
I can't do anything better on the antivirus side. 403 with message "Forbidden" is neither too much descriptive. :/

**Comment by VicDeo** [comment]
grr. It drive me nuts. All methods of Config mock return `null` on the second test execution.

**Comment by ms270169** [comment]
@VicDeo , maybe you have an idea how to better visualize the message. Because the information about a detected virus is inside the HTTP 415 response. If the message of the response would be shown, everything would be clear for the user...
```
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>OCA\DAV\Connector\Sabre\Exception\UnsupportedMediaType</s:exception>
  <s:message>Virus Eicar-Test-Signature is detected in the file. Upload cannot be completed.</s:message>
</d:error>
```

**Comment by VicDeo** [comment]
I still have no idea why PHPUnit mocks return null for the second testcase. 
I created own mocks instead of mocking with PHPUnit API.
It works this way.

**Comment by VicDeo** [comment]
@PVince81 ready from my POV

**Comment by PVince81** [comment]
@VicDeo from what I see this was tested and only needs a code review, correct ?

**Comment by VicDeo** [comment]
@PVince81 yes
 @ms270169  helped me a lot with a design overview.

**Comment by PVince81** [comment]
The code looks good, great to have tests!

@VicDeo I didn't find where "AVMaxFileSize" is covered / passed in the tests, mind pointing me there ? I guess it probably has another name

**Comment by VicDeo** [comment]
@PVince81 Not covered.
 It is used in background scanner only, because file size is unpredictable while using a stream wrapper.
`BackgroundScanner::getFilesForScan` is the only place in code where is it used.

**Comment by PVince81** [comment]
Ok, fine 👍

**Comment by VicDeo** [comment]
Stable9.1: https://github.com/owncloud/files_antivirus/issues/140

**Comment by zubanst** [comment]
Any back-port plan for 9.0.x? In 9.0.6 I have the same issue described in #125 .

**Comment by VicDeo** [comment]
@zubanst would you like to test https://github.com/owncloud/files_antivirus/pull/144 ?

**Comment by zubanst** [comment]
@VicDeo I would do if I had all the details how to.

**Comment by SergioBertolinSG** [comment]
Updated steps:

```
To reproduce:

1. set StreamMaxLength in /etc/clamd.conf to e.g. 10M
2. Configure the app in daemon or daemon(socket) mode
3. In files_antivirus config in administration change the values:
   Stream Length   to 10000000 bytes
   File size limit, -1 means no limit  to 10000000 bytes
4. cat several 100Mb files with EICAR signature
5. Try upload or background scan
```

Expected: It detects the virus. 
Not expected: It uploads the file.

**Comment by VicDeo** [comment]
@SergioBertolinSG 
Indeed `Stream Length` in files_antivirus configuration should match Clamav  `StreamMaxLength` option.
`File size limit` has no effect on upload.

**Pull request merged**